source/fundamentals/typescript.txt

```diff
@@ -0,0 +1,97 @@
+==========
+TypeScript
+==========
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecols
+
+Overview
+--------
+
+In this guide, you can learn about the **TypeScript** specific features and limitiations
+of the MongoDB Node.js driver. TypeScript is a strongly typed programming
+language that compiles to JavaScript.
+
+All TypeScript features of the driver are optional, and all valid JavaScript
+code written with the driver is also valid TypeScript code. 
+
+For more information on TypeScript, see 
+`the official TypeScript website <https://www.typescriptlang.org/>`__.
+
+Features
+--------
+
+If you use TypeScript, you can specify a type for some classes in the driver.
+All classes that accept a type paramater in the driver have the default type
+:node-api-4.0:`Document <interfaces/document.html>`.
+
+The following classes accept any type that extends the ``Document``
+interface:
+
+- :node-api-4.0:`Collection <classes/collection.html>`
+- :node-api-4.0:`ChangeStream <classes/changestream.html>`
+
+You can pass a type parameter that extends the ``Document`` interface like this:
+
+.. literalinclude:: /code-snippets/typescript/extend-document.ts
+   :language: typescript
+   :linenos:
+
+The following classes accept any type parameter:
+
+- :node-api-4.0:`FindCursor <classes/findcursor.html>`
+- :node-api-4.0:`AggregationCursor <classes/aggregationcursor.html>`
+
+You can find an example that shows how to specify a type for the ``FindCursor``
+class in our 
+:ref:`Find a Document Usage Example <node-driver-findone-usage-example-code-snippet>`.
+
+Limitations
+-----------
+
+.. _node-driver-typescript-limitations-dot-notation:
+
+The driver cannot infer the type of values with keys containing **dot
+notation**. Dot notation is a property access syntax for navigating BSON objects.
+Click on the tabs to see code snippets that highlight this behavior:
+
+.. tabs::
+
+  .. tab:: Dot Notation
+     :tabid: dot-notation
+  
+     The following code snippet does not raise a type error:
+
+     .. literalinclude:: /code-snippets/typescript/dot-notation.ts
+        :language: typescript
+        :linenos:
+        :start-after: start-no-error
+        :end-before: end-no-error
+
+  .. tab:: Nested Objects
+     :tabid: nested-objects
+    
+     The following code snippet raises a type error:
+
+     .. literalinclude:: /code-snippets/typescript/dot-notation.ts
+        :language: typescript
+        :linenos:
+        :start-after: start-error
+        :end-before: end-error
+
+     This is the error:
+
+     .. code-block:: text
+
+        Type 'string' is not assignable to type 'number'.
+
+You should still use dot notation to access nested fields in query and update documents
```

Comment:
Suggestion: "You use dot ... TypeScript. Ensure your nested field values have your intended type."

Reply:
I think the edit loses some meaning from the original sentence. I am trying to say:

- Even though you lose a feature of TypeScript which caused you to use TypeScript when you use dot notation, you should still use dot notation (I found this surprising). 
- You can't ensure through the TS language that your type is correct, so you need to manually check with your eyes and trace the code yourself. There is not a way at present to "ensure", so I think "check" is a better word here.

Tweaked the sentence to try and highlight these points: 

Despite the lack of type safety, you should still use dot notation to access nested
fields in query and update documents when you use TypeScript. You must
manually check that your nested field values have your intended type.